<h1>پیش‌بینی سود فروش محصولات با استفاده از مدل‌های رگرسیونی</h1>

<p dir="rtl">
در این پروژه، با استفاده از داده‌های فروش محصولات، هدف ما طراحی مدلی است که بتواند مقدار سود (<code>Profit</code>) را بر اساس ویژگی‌های مختلف سفارش، محصول، و زمان پیش‌بینی کند. این مدل می‌تواند به تصمیم‌گیری‌های استراتژیک شرکت‌ها در زمینه فروش، بازاریابی و مدیریت موجودی کمک کند.
</p>


<h2>وارد کردن کتابخانه‌ها و مدل‌ها</h2>

<p dir="rtl">
در این بخش، کتابخانه‌های مورد نیاز برای پیش‌پردازش داده‌ها، تقسیم داده‌ها، آموزش مدل‌های رگرسیونی، و ارزیابی عملکرد مدل‌ها وارد شده‌اند.
</p>



In [272]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, RobustScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge, ElasticNet
from sklearn.svm import SVR
from sklearn.ensemble import GradientBoostingRegressor, RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error, mean_absolute_percentage_error
import pickle

<h2>بارگذاری و بررسی اولیه داده‌ها</h2>

<p dir="rtl">
در این مرحله، داده‌ها از فایل <code>phase3_profit_prediction_dataset.csv</code> بارگذاری شده و ساختار کلی آن بررسی می‌شود. این بررسی شامل مشاهده نمونه‌ای از داده‌ها، نوع داده‌ها در هر ستون، و تعداد مقادیر یکتای هر ستون است.
</p>

<p dir="rtl">
هدف از این مرحله، آشنایی اولیه با ساختار داده‌ها، شناسایی ستون‌های متنی، عددی، و بررسی وجود مقادیر پرت یا تکراری است. این اطلاعات در طراحی مراحل بعدی پیش‌پردازش و انتخاب مدل بسیار مؤثر خواهد بود.
</p>


In [273]:
df = pd.read_csv('phase3_profit_prediction_dataset.csv')

In [274]:
df.head()

,Profit,Discount,ProfitMargin,Quantity,Sales,SalesAfterDiscount,Shipping Cost,Ship Mode,Category,CategorySubcategory,...,Order Priority,Region,Country,City,DayOfWeek,IsWeekend,Day,Month,Year,Segment
0,-6599.979980,0.7,'-1.466668743,5,4499.979980,1349.994048,451.630005,Standard Class,Technology,Technology - Machines,...,Low,East,United States,Lancaster,3,False,26,11,2013,Consumer
1,-3839.989990,0.5,'-0.47999995,4,7999.979980,3999.989990,674.820007,Same Day,Technology,Technology - Machines,...,High,South,United States,Burlington,4,False,5,11,2014,Corporate
2,-3701.889893,0.8,'-1.700001727,8,2177.580078,435.515990,172.070007,Standard Class,Office Supplies,Office Supplies - Binders,...,Medium,Central,United States,San Antonio,3,False,26,7,2011,Consumer
3,-3399.979980,0.7,'-1.333330716,5,2549.989990,764.997027,120.309998,Standard Class,Technology,Technology - Machines,...,Medium,West,United States,Louisville,6,True,18,4,2014,Home Office
4,-3059.820068,0.5,'-0.900037106,12,3399.659912,1699.829956,286.869995,Standard Class,Office Supplies,Office Supplies - Appliances,...,Medium,South,Portugal,Lisbon,2,False,10,6,2013,Corporate


In [275]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Profit               30000 non-null  float64
 1   Discount             30000 non-null  float64
 2   ProfitMargin         30000 non-null  object 
 3   Quantity             30000 non-null  int64  
 4   Sales                30000 non-null  float64
 5   SalesAfterDiscount   30000 non-null  float64
 6   Shipping Cost        30000 non-null  float64
 7   Ship Mode            30000 non-null  object 
 8   Category             30000 non-null  object 
 9   CategorySubcategory  30000 non-null  object 
 10  Product Name         30000 non-null  object 
 11  IsReturned           30000 non-null  object 
 12  Market               30000 non-null  object 
 13  Order Priority       30000 non-null  object 
 14  Region               30000 non-null  object 
 15  Country              30000 non-null 

In [276]:
df.nunique().sort_values()

IsReturned                 2
IsWeekend                  2
Segment                    3
Category                   3
Ship Mode                  4
Year                       4
Order Priority             4
Market                     7
DayOfWeek                  7
Month                     12
Region                    13
Quantity                  14
CategorySubcategory       17
Discount                  27
Day                       31
Country                  142
City                    3197
Product Name            3413
Shipping Cost           9235
Sales                  13674
Profit                 14076
ProfitMargin           15925
SalesAfterDiscount     16593
dtype: int64

<h2>تقسیم داده‌ها به آموزش و تست</h2>

<p dir="rtl">
در این مرحله، کل داده‌ها به دو بخش آموزش (۸۰٪) و تست (۲۰٪) تقسیم می‌شوند تا بتوان مراحل پیش‌پردازش را به‌صورت جداگانه روی هر بخش انجام داد. در این مرحله، ستون هدف (<code>Profit</code>) هنوز جدا نشده است و در مراحل بعدی برای ساخت ویژگی‌های آماری و مشتق‌شده از آن استفاده خواهد شد.
</p>

<p dir="rtl">
جدا کردن ویژگی‌ها (<code>X</code>) و هدف (<code>y</code>) پس از پیش‌پردازش انجام خواهد شد تا از نشت اطلاعات در مدل‌سازی جلوگیری شود.
</p>

<p dir="rtl">
پارامترهای استفاده‌شده:
</p>
<ul dir="rtl">
  <li><code>test_size=0.2</code>: ۲۰٪ از داده‌ها برای تست کنار گذاشته می‌شود</li>
  <li><code>random_state=42</code>: برای تکرارپذیری نتایج، یک مقدار ثابت برای تصادفی‌سازی استفاده شده است</li>
</ul>



In [277]:
train_data, test_data = train_test_split(df, test_size=0.2, random_state=42)

<h2>پیش‌پردازش داده‌ها و آماده‌سازی برای مدل‌سازی</h2>

<p dir="rtl">
در این مرحله، داده‌ها برای آموزش مدل‌های رگرسیونی آماده‌سازی می‌شوند. این شامل پاک‌سازی مقادیر رشته‌ای، تبدیل ویژگی‌های متنی به عددی، حذف ستون‌های غیرمفید، ساخت ویژگی‌های ترکیبی، و نرمال‌سازی ویژگی‌های عددی است.
</p>

<p dir="rtl">
مراحل انجام‌شده:
</p>
<ul dir="rtl">
  <li>حذف ستون <code>Product Name</code> به‌دلیل یکتایی بالا و بی‌اثر بودن در مدل</li>
  <li>حذف ستون <code>ProfitMargin</code> به‌دلیل وابستگی مستقیم به هدف مدل و جلوگیری از نشت اطلاعات</li>
  <li>تبدیل ستون <code>IsReturned</code> به مقدار عددی (1 برای Yes، 0 برای No)</li>
  <li>تبدیل ستون‌های دسته‌ای با تعداد کلاس کم به One-Hot Encoding</li>
  <li>کاهش کلاس‌های زیاد در ستون‌های <code>Country</code> و <code>City</code> به ۱۰ مقدار پرتکرار و تبدیل بقیه به "Other"</li>
  <li>ساخت ویژگی‌های ترکیبی مانند <code>DiscountRate</code>، <code>ShippingRatio</code>، <code>WeekendImpact</code>، <code>EffectiveDiscount</code> و <code>NetSalesRatio</code></li>
  <li>نرمال‌سازی ویژگی‌های عددی با <code>RobustScaler</code> برای کاهش تأثیر داده‌های پرت</li>
</ul>


In [278]:
train_data = train_data.drop(columns=["Product Name"])
train_data = train_data.drop(columns=["ProfitMargin"])

print(train_data.columns.tolist())

['Profit', 'Discount', 'Quantity', 'Sales', 'SalesAfterDiscount', 'Shipping Cost', 'Ship Mode', 'Category', 'CategorySubcategory', 'IsReturned', 'Market', 'Order Priority', 'Region', 'Country', 'City', 'DayOfWeek', 'IsWeekend', 'Day', 'Month', 'Year', 'Segment']


In [279]:
train_data['IsReturned'] = train_data['IsReturned'].map({'Yes': 1, 'No': 0})
train_data['IsReturned'].info()

<class 'pandas.core.series.Series'>
Index: 24000 entries, 21753 to 23654
Series name: IsReturned
Non-Null Count  Dtype
--------------  -----
24000 non-null  int64
dtypes: int64(1)
memory usage: 375.0 KB


In [280]:
low_cardinality = ['Ship Mode', 'Category', 'CategorySubcategory', 'Market', 'Order Priority', 'Region', 'Segment']
train_data = pd.get_dummies(train_data, columns=low_cardinality, drop_first=True)

In [281]:
top_countries = train_data['Country'].value_counts().nlargest(10).index.tolist()
top_cities = train_data['City'].value_counts().nlargest(10).index.tolist()
train_data[Country] = train_data[Country].apply(lambda x: x if x in top_countries else 'Other')
train_data[City] = train_data[City].apply(lambda x: x if x in top_cities else 'Other')

train_data = pd.get_dummies(train_data, columns=['Country', 'City'], drop_first=True)

In [282]:
# نسبت تخفیف به فروش
train_data['DiscountRate'] = train_data['Discount'] / train_data['Sales']

# نسبت هزینه ارسال به فروش
train_data['ShippingRatio'] = train_data['Shipping Cost'] / train_data['Sales']

# تأثیر آخر هفته روی هزینه ارسال
train_data['WeekendImpact'] = train_data['IsWeekend'].astype(int) * train_data['Shipping Cost']

# تخفیف مؤثر روی قیمت نهایی
train_data['EffectiveDiscount'] = train_data['Sales'] - train_data['SalesAfterDiscount']

# نسبت فروش پس از تخفیف به فروش اصلی
train_data['NetSalesRatio'] = train_data['SalesAfterDiscount'] / train_data['Sales']


In [283]:
num_cols = ['Discount', 'Quantity', 'Sales', 'SalesAfterDiscount', 'Shipping Cost', 'DayOfWeek','DiscountRate','ShippingRatio', 'WeekendImpact','EffectiveDiscount', 'NetSalesRatio', 'Day',
            'Month', 'Year']

scaler = RobustScaler()
train_data[num_cols] = scaler.fit_transform(train_data[num_cols])

In [284]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24000 entries, 21753 to 23654
Data columns (total 81 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Profit                                            24000 non-null  float64
 1   Discount                                          24000 non-null  float64
 2   Quantity                                          24000 non-null  float64
 3   Sales                                             24000 non-null  float64
 4   SalesAfterDiscount                                24000 non-null  float64
 5   Shipping Cost                                     24000 non-null  float64
 6   IsReturned                                        24000 non-null  int64  
 7   DayOfWeek                                         24000 non-null  float64
 8   IsWeekend                                         24000 non-null  bool   
 9   Day               

<h2>جداسازی ویژگی‌ها و هدف از داده آموزش</h2>

<p dir="rtl">
در این مرحله، ستون هدف (<code>Profit</code>) از داده آموزش جدا می‌شود تا به‌عنوان خروجی مدل استفاده شود. سپس این ستون از مجموعه ویژگی‌ها (<code>X_train</code>) حذف می‌شود تا از نشت اطلاعات در مدل‌سازی جلوگیری شود.
</p>




In [285]:
y_train = train_data['Profit']
X_train = train_data.drop(columns=['Profit'])


<h2>تعریف تابع آماده‌سازی داده جدید برای پیش‌بینی</h2>

<p dir="rtl">
برای استفاده عملی از مدل نهایی، تابعی تعریف شده است که داده‌های جدید را دقیقاً مطابق مراحل پیش‌پردازش انجام‌شده روی داده‌های آموزش، آماده‌سازی می‌کند. این تابع شامل حذف ستون‌های غیرضروری، تبدیل ویژگی‌های متنی به عددی، کاهش کلاس‌های زیاد، ساخت ویژگی‌های ترکیبی، و نرمال‌سازی ویژگی‌های عددی با استفاده از <code>RobustScaler</code> آموزش‌دیده است.
</p>

<p dir="rtl">
در پایان، این تابع ستون هدف (<code>Profit</code>) را جدا کرده و دو خروجی <code>X</code> و <code>y</code> بازمی‌گرداند تا مستقیماً وارد مدل شوند.
</p>

In [286]:

def preprocess_input_data(new_df, scaler, scaler_cols, top_countries, top_cities):
    if 'Product Name' in new_df.columns:
        new_df = new_df.drop(columns=['Product Name'])
    if 'ProfitMargin' in new_df.columns:
        new_df = new_df.drop(columns=['ProfitMargin'])
    new_df['IsReturned'] = new_df['IsReturned'].map({'Yes': 1, 'No': 0})
    new_df['IsWeekend'] = new_df['IsWeekend'].astype(int)

    new_df['DiscountRate'] = new_df['Discount'] / new_df['Sales']
    new_df['ShippingRatio'] = new_df['Shipping Cost'] / new_df['Sales']
    new_df['WeekendImpact'] = new_df['IsWeekend'] * new_df['Shipping Cost']
    new_df['EffectiveDiscount'] = new_df['Sales'] - new_df['SalesAfterDiscount']
    new_df['NetSalesRatio'] = new_df['SalesAfterDiscount'] / new_df['Sales']


    low_cardinality = ['Ship Mode', 'Category', 'CategorySubcategory', 'Market', 'Order Priority', 'Region', 'Segment']
    new_df = pd.get_dummies(new_df, columns=low_cardinality, drop_first=True)

    for col, top_values in zip(['Country', 'City'], [top_countries, top_cities]):
        new_df[col] = new_df[col].apply(lambda x: x if x in top_values else 'Other')
    new_df = pd.get_dummies(new_df, columns=['Country', 'City'], drop_first=True)

    new_df[scaler_cols] = scaler.transform(new_df[scaler_cols])

    y = new_df['Profit']
    X = new_df.drop(columns=['Profit'])

    return X, y


In [287]:
X_test, y_test = preprocess_input_data(
    test_data,
    scaler,
    num_cols,
    top_countries,
    top_cities
)


<h2>تعریف مدل‌ها و تنظیمات هایپرپارامترها</h2>

<p dir="rtl">
در این مرحله، مجموعه‌ای از مدل‌های رگرسیونی تعریف شده‌اند که هرکدام با مجموعه‌ای از هایپرپارامترها برای تنظیم بهینه عملکردشان آماده شده‌اند. این مدل‌ها در مرحله بعد با استفاده از <code>GridSearchCV</code> آموزش داده خواهند شد.
</p>

<p dir="rtl">
هر مدل دارای مجموعه‌ای از پارامترهاست که در مرحله تنظیم با <code>GridSearchCV</code> بررسی می‌شوند تا بهترین ترکیب برای بیشترین دقت انتخاب شود.
</p>


In [288]:

models = {
    'Ridge': {
        'model': Ridge(),
        'params': {
            'alpha': [0.01, 0.1, 1, 10, 100]
        }
    },
    'ElasticNet': {
        'model': ElasticNet(),
        'params': {
            'alpha': [0.01, 0.1, 1, 10],
            'l1_ratio': [0.1, 0.5, 0.9],
            'max_iter': [1000, 5000],
            'tol': [1e-3, 1e-4]
        }
    },
    'SVR': {
        'model': SVR(),
        'params': {
            'C': [0.1, 1, 10],
            'kernel': ['linear', 'rbf'],
            'epsilon': [0.01, 0.1]
        }
    },
    'GradientBoosting': {
        'model': GradientBoostingRegressor(),
        'params': {
            'n_estimators': [100, 200],
            'learning_rate': [0.01, 0.1],
            'max_depth': [3, 5]
        }
    },
    'RandomForest': {
        'model': RandomForestRegressor(),
        'params': {
            'n_estimators': [100, 200],
            'max_depth': [5, 10, None],
            'min_samples_split': [2, 5]
        }
    },
    'XGBoost': {
        'model': XGBRegressor(),
        'params': {
            'n_estimators': [100, 200],
            'learning_rate': [0.01, 0.1],
            'max_depth': [3, 5],
            'reg_alpha': [0, 0.1],
            'reg_lambda': [1, 5]
        }
    }
}

<h2>آموزش مدل‌ها با GridSearchCV و ارزیابی روی داده‌های تست</h2>

<p dir="rtl">
در این مرحله، با استفاده از <code>GridSearchCV</code>، بهترین ترکیب هایپرپارامترها برای هر مدل پیدا می‌شود. سپس مدل نهایی روی داده‌های تست ارزیابی می‌شود تا عملکرد واقعی آن سنجیده شود.
</p>
<p dir="rtl">
مراحل انجام‌شده:
</p>
<ul dir="rtl">
  <li>استفاده از <code>GridSearchCV</code> برای جستجوی بهترین ترکیب هایپرپارامترها با معیار <code>R²</code></li>
  <li>استفاده از <code>cv=2</code> برای اعتبارسنجی متقاطع سریع</li>
  <li>ذخیره بهترین مدل، امتیاز و پارامترها در دیکشنری <code>best_models</code></li>
  <li>پیش‌بینی روی داده‌های تست و محاسبه <code>R²</code> نهایی برای هر مدل</li>
</ul>


In [289]:
best_models = {}

for name, config in models.items():
    grid = GridSearchCV(config['model'], config['params'], cv=2, scoring='r2', n_jobs=-1)
    grid.fit(X_train, y_train)
    best_models[name] = {
        'best_estimator': grid.best_estimator_,
        'best_score': grid.best_score_,
        'best_params': grid.best_params_
    }
    print(f"{name} - بهترین r2: {grid.best_score_:.4f} با پارامترها: {grid.best_params_}")

Ridge - بهترین r2: 0.5573 با پارامترها: {'alpha': 0.01}
ElasticNet - بهترین r2: 0.5568 با پارامترها: {'alpha': 0.01, 'l1_ratio': 0.9, 'max_iter': 5000, 'tol': 0.001}
SVR - بهترین r2: 0.5472 با پارامترها: {'C': 0.1, 'epsilon': 0.01, 'kernel': 'linear'}
GradientBoosting - بهترین r2: 0.7670 با پارامترها: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100}
RandomForest - بهترین r2: 0.7679 با پارامترها: {'max_depth': 5, 'min_samples_split': 2, 'n_estimators': 200}
XGBoost - بهترین r2: 0.7299 با پارامترها: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100, 'reg_alpha': 0.1, 'reg_lambda': 1}


In [290]:
X_test = X_test.reindex(columns=X_train.columns, fill_value=0)

print("\nارزیابی روی test set:")
for name, model_info in best_models.items():
    test_pred = model_info['best_estimator'].predict(X_test)
    test_r2 = r2_score(y_test, test_pred)
    print(f"{name} - R² روی test: {test_r2:.4f}")


ارزیابی روی test set:
Ridge - R² روی test: 0.5637
ElasticNet - R² روی test: 0.5649
SVR - R² روی test: 0.6839
GradientBoosting - R² روی test: 0.8209
RandomForest - R² روی test: 0.7670
XGBoost - R² روی test: 0.7962


<h2>انتخاب مدل نهایی</h2>

<p dir="rtl">
پس از آموزش و ارزیابی چندین مدل رگرسیونی شامل <strong>Ridge</strong>، <strong>ElasticNet</strong>، <strong>SVR</strong>، <strong>GradientBoosting</strong>، <strong>RandomForest</strong> و <strong>XGBoost</strong>، در نهایت مدل <strong>GradientBoostingRegressor</strong> به‌عنوان مدل نهایی انتخاب شد.
</p>

<h3>دلایل انتخاب:</h3>
<ul dir="rtl">
  <li><strong>دقت بالا:</strong> مقدار R² روی داده‌های تست برابر با <code>0.8209</code> بود که نشان‌دهنده قدرت پیش‌بینی بسیار خوب مدل است.</li>
  <li><strong>تعادل بین دقت و پیچیدگی:</strong> در مقایسه با مدل‌های RandomForest و XGBoost که دقت مشابهی داشتند، GradientBoosting سبک‌تر و سریع‌تر است.</li>
  <li><strong>قابلیت تعمیم‌پذیری:</strong> مدل روی داده‌های تست عملکرد پایداری داشت و دچار overfitting نشد.</li>
  <li><strong>قابل تفسیر بودن:</strong> امکان استخراج اهمیت ویژگی‌ها و تحلیل رفتار مدل وجود دارد.</li>
</ul>

<p dir="rtl">
در نتیجه، مدل <strong>GradientBoostingRegressor</strong> با پارامترهای بهینه‌شده زیر به‌عنوان مدل نهایی انتخاب شد:
</p>

<pre>
GradientBoostingRegressor(
    learning_rate=0.1,
    max_depth=3,
    n_estimators=100
)
</pre>


<h2>ارزیابی نهایی مدل GradientBoosting</h2>

<p dir="rtl">
پس از انتخاب مدل نهایی (GradientBoostingRegressor)، عملکرد آن روی داده‌های تست با استفاده از متریک‌های کلیدی ارزیابی می‌شود.
</p>

<p dir="rtl">
متریک‌های استفاده‌شده:
</p>
<ul dir="rtl">
  <li><strong>R² (ضریب تعیین):</strong> نشان می‌دهد چه درصدی از واریانس داده‌ها توسط مدل توضیح داده شده است. هرچه به ۱ نزدیک‌تر، بهتر.</li>
  <li><strong>MAE (میانگین قدر مطلق خطا):</strong> میانگین فاصله پیش‌بینی‌ها از مقادیر واقعی. قابل تفسیر و مقاوم به داده‌های پرت.</li>
  <li><strong>MSE (میانگین مربعات خطا):</strong> میانگین مربع خطاها. به خطاهای بزرگ حساس‌تر است و برای بررسی دقت کلی مدل مفید است.</li>
</ul>


In [291]:

final_model = best_models['GradientBoosting']['best_estimator']
y_pred = final_model.predict(X_test)

print("ارزیابی نهایی مدل GradientBoosting:")
print("R²:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))
print("MSE:", mean_squared_error(y_test, y_pred))





ارزیابی نهایی مدل GradientBoosting:
R²: 0.8209005479711321
MAE: 14.835253199982855
MSE: 1980.696517442462


<h2>تحلیل اهمیت ویژگی‌ها (Feature Importance)</h2>

<p dir="rtl">
مدل GradientBoostingRegressor این قابلیت را دارد که اهمیت نسبی هر ویژگی را در پیش‌بینی خروجی محاسبه کند. این تحلیل به ما کمک می‌کند تا بفهمیم کدام ویژگی‌ها بیشترین تأثیر را در پیش‌بینی سود داشته‌اند.
</p>

<p dir="rtl">
در این کد:
</p>
<ul dir="rtl">
  <li>ویژگی‌ها و میزان اهمیت آن‌ها از مدل استخراج می‌شوند.</li>
  <li>با استفاده از <code>DataFrame</code> مرتب‌سازی انجام می‌شود تا مهم‌ترین ویژگی‌ها در صدر قرار گیرند.</li>
  <li>در نهایت، ۲۰ ویژگی برتر چاپ می‌شوند که بیشترین نقش را در پیش‌بینی سود ایفا کرده‌اند.</li>
</ul>

<p dir="rtl">
این تحلیل می‌تواند برای تصمیم‌گیری‌های تجاری، انتخاب ویژگی‌های کلیدی، و بهینه‌سازی مدل‌های آینده بسیار مفید باشد.
</p>


In [295]:
importances = final_model.feature_importances_
features = X_test.columns

importance_df = pd.DataFrame({'features': features, 'importances': importances})
importance_df = importance_df.sort_values(by='importances', ascending=False)
print(importance_df.head(20))


                                         features  importances
78                              EffectiveDiscount     0.797771
0                                        Discount     0.075447
79                                  NetSalesRatio     0.043883
39                             Order Priority_Low     0.016302
45                                    Region_East     0.012929
76                                  ShippingRatio     0.010449
1                                        Quantity     0.008233
6                                       DayOfWeek     0.007220
2                                           Sales     0.005806
3                              SalesAfterDiscount     0.002935
4                                   Shipping Cost     0.002796
9                                           Month     0.002775
75                                   DiscountRate     0.002002
10                                           Year     0.001297
12                         Ship Mode_Second Class     0

<h2>تحلیل اهمیت ویژگی‌ها در مدل نهایی</h2>

<p dir="rtl">
مدل نهایی نشان می‌دهد که برخی ویژگی‌ها تأثیر بسیار بیشتری در پیش‌بینی سود (<code>Profit</code>) دارند. در ادامه، مهم‌ترین نکات تحلیلی آورده شده است:
</p>

<h3>ویژگی‌های بسیار مؤثر</h3>
<ul dir="rtl">
  <li><strong>EffectiveDiscount</strong> با ضریب اهمیت <code>0.797</code>، مهم‌ترین عامل در پیش‌بینی سود است. این ویژگی نشان‌دهنده اختلاف بین قیمت اصلی و قیمت بعد از تخفیف است، که مستقیماً بر سود نهایی اثر می‌گذارد.</li>
  <li><strong>Discount</strong> و <strong>NetSalesRatio</strong> نیز نقش قابل توجهی دارند. تخفیف مستقیم و نسبت فروش خالص به فروش اولیه، هر دو نشان‌دهنده سیاست‌های قیمت‌گذاری و تأثیر آن‌ها بر سود هستند.</li>
</ul>

<h3>ویژگی‌های عملیاتی با اهمیت متوسط</h3>
<ul dir="rtl">
  <li>ویژگی‌هایی مثل <code>Order Priority_Low</code>، <code>Region_East</code>، و <code>ShippingRatio</code> اهمیت متوسطی دارند.</li>
  <li>این‌ها به نحوه ارسال، منطقه جغرافیایی، و اولویت سفارش مربوط می‌شوند که می‌توانند بر هزینه‌ها و زمان تحویل تأثیر بگذارند.</li>
</ul>

<h3>ویژگی‌های کم‌اثر</h3>
<ul dir="rtl">
  <li>ویژگی‌هایی مثل <code>Month</code>، <code>Year</code>، <code>Ship Mode</code>، و برخی دسته‌بندی‌های محصول، اهمیت بسیار کمی دارند.</li>
  <li>این نشان می‌دهد که زمان سفارش یا نوع حمل‌ونقل تأثیر مستقیمی بر سود ندارد یا تأثیر آن‌ها در داده فعلی کم بوده است.</li>
</ul>




<h2>ذخیره‌سازی مدل نهایی با pickle</h2>

<p dir="rtl">
برای استفاده مجدد از مدل آموزش‌دیده بدون نیاز به آموزش دوباره، آن را با استفاده از کتابخانه <code>pickle</code> در قالب فایل باینری ذخیره می‌کنیم.
</p>

<p dir="rtl">
در این کد:
</p>
<ul dir="rtl">
  <li><code>'gradient_boosting_profit_model.pkl'</code> نام فایلی است که مدل در آن ذخیره می‌شود. می‌توان این نام را به دلخواه تغییر داد.</li>
  <li><code>'wb'</code> به معنای "نوشتن باینری" است؛ یعنی فایل به‌صورت باینری باز می‌شود تا داده‌ها در آن ذخیره شوند.</li>
  <li><code>pickle.dump(...)</code> مدل آموزش‌دیده را در فایل مشخص‌شده ذخیره می‌کند.</li>
</ul>

<p dir="rtl">
این فایل را می‌توان در آینده بارگذاری کرد و مستقیماً برای پیش‌بینی استفاده نمود، بدون نیاز به آموزش مجدد مدل.البته باید داده ها با تابع بالا آماده بشن بعد به مدل داده بشن
</p>


In [294]:
with open('gradient_boosting_profit_model.pkl', 'wb') as f:
    pickle.dump(final_model, f)